File: prob2.java
public class DiagonalOrder {
    public static int[] findDiagonalOrder(int[][] matrix) {
        // Check for empty matrix
        if (matrix == null || matrix.length == 0) {
            return new int[0];
        }

        // Get number of rows and columns
        int m = matrix.length;
        int n = matrix[0].length;

        // Create result array to store diagonal elements
        int[] result = new int[m * n];

        // Initialize variables for current row, column, and direction
        int row = 0, col = 0, d = 1;

        // Iterate over all elements in the matrix
        for (int i = 0; i < m * n; i++) {
            // Add current element to result array
            result[i] = matrix[row][col];

            // Move to next diagonal element
            row -= d;
            col += d;

            // If we go out of bounds, change direction and adjust row/column
            if (row >= m) {
                row = m - 1;
                col += 2;
                d = -d;
            }
            if (col >= n) {
                col = n - 1;
                row += 2;
                d = -d;
            }
            if (row < 0) {
                row = 0;
                d = -d;
            }
            if (col < 0) {
                col = 0;
                d = -d;
            }
        }

        // Return the result array
        return result;
    }
}

File: prob2.py
class Solution:
    def findDiagonalOrder(self, matrix: List[List[int]]) -> List[int]:
        # Check for empty matrix
        if not matrix or not matrix[0]:
            return []

        # Get number of rows and columns
        m, n = len(matrix), len(matrix[0])

        # Create result array to store diagonal elements
        result = []

        # Initialize variables for current row, column, and direction
        row, col, direction = 0, 0, 1

        # Iterate over all elements in the matrix
        for i in range(m * n):
            # Add current element to result array
            result.append(matrix[row][col])

            # Move to next diagonal element
            row -= direction
            col += direction

            # If we go out of bounds, change direction and adjust row/column
            if row >= m:
                row = m - 1
                col += 2
                direction = -direction
            if col >= n:
                col = n - 1
                row += 2
                direction = -direction
            if row < 0:
                row = 0
                direction = -direction
            if col < 0:
                col = 0
                direction = -direction

        # Return the result array
        return result
